feat: AttachableBehaviour and ComponentController

NoelStephensUnity opened the pull request:
## Depends on
This PR depends upon the fix for the `SendTo.NotMe` fix in #3551.

## AttachableBehaviour and Support Components

The purpose of this PR (feat) is to address the complexity of "picking up" or "dropping" an item in the world which can become complex when using the traditional `NetworkObject` parenting pattern.  In this PR there are three primary components added to help reduce this complexity:

- `AttachableBehaviour`: Provides "out of the box" support for attaching (i.e. parenting) a nested child `GameObject` that includes an `AttachableBehaviour` component to another nested child `GameObject` with an `AttachableNode` component that is associated with a different `NetworkObject`.
- `AttachableNode`: This component is required by the `AttachableBehaviour` component in order to be able to attach (i.e. parent) to another GameObject without having to parent the entire `NetworkObject` component the `AttachableBehaviour` component is associated with.
- `ComponentController`: This component provides users the ability to synchronize the enabling or disabling of any `Object` derived component that has an `enabled` property.

_This PR also incorporates a new "Helpers" subfolder under the NGO components folder where additional helper components will live._ 

## Documentation Update

## New Documentation
[AttachableBehaviour](https://github.com/Unity-Technologies/com.unity.netcode.gameobjects/blob/feat/attachable-networkbehaviour-and-object-controller/com.unity.netcode.gameobjects/Documentation~/components/Helpers/attachablebehaviour.md)

### Network Components Section Update
<img width="957" height="589" alt="image" src="https://github.com/user-attachments/assets/48e9edaf-cf8d-4a61-b0b9-c0369bb8f26f" />

### New Foundational Components Section
<img width="1145" height="831" alt="image" src="https://github.com/user-attachments/assets/f3dc71b4-cd07-4884-a5a6-fc7c78e9e24c" />

### New Helper Components Section
<img width="1015" height="537" alt="image" src="https://github.com/user-attachments/assets/407636b8-9085-4b80-8bb4-d55b18512ce1" />

## NetworkBehaviour.OnNetworkPreDespawn

Added another virtual method to `NetworkBehaviour`, `OnNetworkPreDespawn`, that is invoked before running through the despawn sequence for the `NetworkObject` and all `NetworkBehaviour` children of the `NetworkObject` being despawned. This provides an opportunity to do any kind of cleanup up or last micro-second state updates prior to despawning.

<!-- Add short version of the JIRA ticket to the PR title (e.g. "feat: new shiny feature [MTT-123]") -->

## Changelog

- Added: `AttachableBehaviour` helper component to provide an alternate approach to parenting items without using the `NetworkObject` parenting.
- Added : `AttachableNode` helper component that is used by `AttachableBehaviour` as the target node for parenting.
- Added: `ComponentController` helper component that can be used to synchronize the enabling and disabling of components and can be used in conjunction with `AttachableBehaviour`.
- Added: `NetworkBehaviour.OnNetworkPreDespawn` that is invoked before running through the despawn sequence for the `NetworkObject` and all `NetworkBehaviour` children of the `NetworkObject` being despawned.


<!--  Uncomment and mark items off with a * if this PR deprecates any API:
### Deprecated API
- [ ] An `[Obsolete]` attribute was added along with a `(RemovedAfter yyyy-mm-dd)` entry.
- [ ] An [api updater] was added.
- [ ] Deprecation of the API is explained in the CHANGELOG.
- [ ] The users can understand why this API was removed and what they should use instead.
-->

## Documentation
[//]: # (
This section is REQUIRED and should mention what documentation changes were following the changes in this PR. 
We should always evaluate if the changes in this PR require any documentation changes.
)

- Includes documentation updates:
  - Added new documentation section.
  - Refactored network components section.
  - Fixed some image links.


## Testing & QA
[//]: #  (
This section is REQUIRED and should describe how the changes were tested and how should they be tested when Playtesting for the release.
It can range from "edge case covered by unit tests" to "manual testing required and new sample was added".
Expectation is that PR creator does some manual testing and provides a summary of it here.)

### Functional Testing
[//]: # (If checked, List manual tests that have been performed.)
_Manual testing :_
- [X] `Manual testing done`
  - Asteroids `AttachableBehaviour` manual test [PR-42](https://github.cds.internal.unity3d.com/unity/Asteroids-CMB-NGO-Sample/pull/42)

_Automated tests:_
- [ ] `Covered by existing automated tests`
- [X] `Covered by new automated tests`

_Does the change require QA team to:_

- [ ] `Review automated tests`?
- [ ] `Execute manual tests`?

If any boxes above are checked, please add QA as a PR reviewer.


## Backport

This is a v2.x only feature.

<!-- If this is a backport:
 - Add the following to the PR title: "\[Backport\] ..." .
 - Link to the original PR.
If this needs a backport - state this here
If a backport is not needed please provide the reason why.
If the "Backports" section is not present it will lead to a CI test failure. 
-->

NoelStephensUnity committed:
update

Adding AttachableBehaviour and ObjectController.

NoelStephensUnity committed:
update

Renaming ObjectController to ComponentController.
Added some additional validation checking and handling.
Updated XML API.

NoelStephensUnity committed:
update

Adding helpers meta.

NoelStephensUnity committed:
Merge branch 'develop-2.0.0' into feat/attachable-networkbehaviour-and-object-controller

NoelStephensUnity committed:
Merge branch 'develop-2.0.0' into feat/attachable-networkbehaviour-and-object-controller

NoelStephensUnity committed:
refactor

Adding an AttachableNode as the target for AttachableBehaviour.
Refactoring AttachableBehaviour.

NoelStephensUnity committed:
test

Adding new test for attachables.

NoelStephensUnity committed:
style

Replacing any improperly spelled "detatch" with "detach".
XML API and private methods.

NoelStephensUnity committed:
style - PVP

Minor XML API fixes.

NoelStephensUnity committed:
style - standards

Simplified the nameof AttachableBehaviour.Detach to just Detach.

NoelStephensUnity committed:
Merge branch 'develop-2.0.0' into feat/attachable-networkbehaviour-and-object-controller

NoelStephensUnity committed:
refactor

Refactoring the ComponentController to provide more flexibility as well as being able to have component entries that will apply the inverse of the current ComponentController's current state....which allows for switching between different sets of components depending upon the controller's state.

NoelStephensUnity committed:
update

Made some minor adjustments while writing the base test for ComponentController.

NoelStephensUnity committed:
test

Adding the base ComponentController test.

NoelStephensUnity committed:
test - update

switching to a Light component as opposed to BoxCollider as BoxCollider requires the physics package and we are just testing the functionality of ComponentController and not specifically any one other type of component.

NoelStephensUnity committed:
style

Removing using directive.

NoelStephensUnity committed:
Merge branch 'develop-2.0.0' into feat/attachable-networkbehaviour-and-object-controller

NoelStephensUnity committed:
Merge branch 'develop-2.0.0' into feat/attachable-networkbehaviour-and-object-controller

NoelStephensUnity committed:
update

Updating the change log entries.

NoelStephensUnity committed:
Merge branch 'develop-2.0.0' into feat/attachable-networkbehaviour-and-object-controller

NoelStephensUnity committed:
update

Removing debug log info.

NoelStephensUnity committed:
update

Work in progress adjustments for usability under various scenarios.

NoelStephensUnity committed:
refactor

Using RPCs and synchronizing the property being set using NetworkBehaviour.OnSynchronize in order to assure order of operations when it comes to messages.

Updated the ComponentController to be able to stagger state updates in the event this occurs (wip and I might remove this part and not allow changes to state until any pending state is finished).

NoelStephensUnity committed:
style

NoelStephensUnity committed:
update & style

Minor adjustment to the range for delays allow it to be zero.
Fixing spelling issues.

NoelStephensUnity committed:
refactor

Had to make some minor adjustments in order to assure that users could handle sending any last micro-second tasks on any spawned instances prior to them despawning.

Added NetworkBehaviour.OnNetworkPreDespawn.
Did a slight order of operations on NetworkManager.Shutdown internal in order to assure sending RPCs during despawn would still work.

Minor adjustments to the new helper components associated with this PR.

NoelStephensUnity committed:
test

Updating the AttachableBehaviourTests to include testing that an attachable will be automatically detatched upon despawning the AttachableNode it is attached to.

NoelStephensUnity committed:
Merge branch 'develop-2.0.0' into feat/attachable-networkbehaviour-and-object-controller

NoelStephensUnity committed:
update

Updating changelog entry.

NoelStephensUnity committed:
Merge branch 'develop-2.0.0' into feat/attachable-networkbehaviour-and-object-controller

NoelStephensUnity committed:
update

Doing a second inspector UI pass to include tool tips and rename each element item to the component's standard inspector view naming where it is the GameObject's name followed by the class name that is separated by capitalization and contained within parenthesis.

NoelStephensUnity committed:
fix

Fixing an exception that can occur when you have a network prefab opened for editing and then you delete the prefab asset before exiting the prefab edit mode.

NoelStephensUnity committed:
style - PVP

Fixing some PVP related issues.

NoelStephensUnity committed:
update

Several improvements on the attach and detach processing.
Added the ability to tie ComponentControllers to AttachableBehaviours in order to auto-notify when something is attaching and detaching.
Several adjustments to fix the issue with ungraceful disconnects and re-synchronizing attachables.
Added forced change based on flags applied, things like when an AttachableNode is despawning, changing ownership, or being destroyed then local instances, whether authority or not, will all force the attach or detach  state.
Added an internal virtual destroy method on NetworkBehaviour to allow for helper components to assure on destroy script is invoked.

NoelStephensUnity committed:
sty;e - PvP

Adding documentation to an undocumented enum value.

NoelStephensUnity committed:
update and test additions

Updating the core components based on some bugs discovered during testing.
Updated the AttachableBehaviourTests to validate the different types of auto-detach flag combinations.
Fixed some spelling issues with "detatch" (not sure why I got into that habit)...corrected to "detach".

NoelStephensUnity committed:
Merge branch 'develop-2.0.0' into feat/attachable-networkbehaviour-and-object-controller

NoelStephensUnity committed:
docs (update,, refactor. and fix)

Adjusting the component section from a file and folder context.
Adjusting the component section for the table of contents.
Adding foundational components and helpers sub-sections to network components section.
Moved foundational components into the foundational (components) folder and updated links and image paths.
Adding AttachableBehaviour document along with images.
Added some place holders for AttachableNode and ComponentController.
Fixing some invalid image paths.

NoelStephensUnity committed:
docs (style-PVP)

Removing white space at the end of lines.

NoelStephensUnity committed:
docs(style-pvp)

Removing trickier white spaces...

EmandM commented on com.unity.netcode.gameobjects/Runtime/Core/NetworkObject.cs:
```diff
@@ -218,7 +218,7 @@ private static void CheckPrefabStage(PrefabStage prefabStage)
                 }
 
-                if (s_PrefabInstance.GlobalObjectIdHash != s_PrefabAsset.GlobalObjectIdHash)
+                if (s_PrefabAsset && s_PrefabInstance && s_PrefabInstance.GlobalObjectIdHash != s_PrefabAsset.GlobalObjectIdHash)
```
This code is inside a block of `if (s_PrefabInstance) {}` (line 200)

```suggestion
                if (s_PrefabAsset && s_PrefabInstance.GlobalObjectIdHash != s_PrefabAsset.GlobalObjectIdHash)
```

NoelStephensUnity replied:
Good catch!
It isn't part of the PR, but that is a small enough adjustment to that area and we shouldn't need any form of test for that either. 
👍

EmandM commented on com.unity.netcode.gameobjects/Runtime/Components/Helpers/AttachableBehaviour.cs:
```diff
@@ -0,0 +1,572 @@
+            /// Used for naming each element entry.
+            /// </summary>
+            [HideInInspector]
+            public string name;
```
Do you mind explaining what this does?

NoelStephensUnity replied:
So, it is an odd thing... in order for the inspector view to name each `AttachableBehaviour.ComponentControllers` item, the class has to have a public `<classname>.name` field.
<img width="688" height="314" alt="image" src="https://github.com/user-attachments/assets/0d7daae9-025d-4787-b66b-715a3ad602a9" />

Which is sort of ironic as that usage pattern breaks the traditional C# pascal casing rule (_which is why a few lines above and below I have to disable the IDE1006 and then just after it re-enable IDE1006_)... of course I could write my own property drawer... but it seemed a bit extreme compared to using what already "just works" in the inspector view.

EmandM commented on com.unity.netcode.gameobjects/Runtime/Components/Helpers/AttachableBehaviour.cs:
```diff
@@ -0,0 +1,572 @@
+        /// as the <see cref="NetworkObject"/>, this will automatically create a child and add an
+        /// <see cref="AttachableBehaviour"/> to that.
+        /// </remarks>
+        protected virtual void OnValidate()
```
To my untrained eye, it looks like there's two `OnValidate` functions and only one will run? What's the difference between this `OnValidate` and the function on line 45?

NoelStephensUnity replied:
The one on line 45 is for `ComponentControllerEntry.OnValidate` which is actually invoked by `AttachableBehaviour.OnValidate` starting around line 97:

```c#
            foreach (var componentController in ComponentControllers)
            {
                if (componentController == null)
                {
                    continue;
                }
                componentController.OnValidate();
            }
```

_(basically extending the validate process for `AttachableBehaviour` to each `ComponentControllerEntry`)_

EmandM commented on com.unity.netcode.gameobjects/Runtime/Components/Helpers/AttachableBehaviour.cs:
```diff
@@ -0,0 +1,572 @@
+            }
+            foreach (var componentController in ComponentControllers)
+            {
+                if (componentController == null)
+                {
+                    continue;
+                }
+                componentController.OnValidate();
```
Definitely a nitpick: [null propagation](https://learn.microsoft.com/en-us/dotnet/csharp/language-reference/operators/member-access-operators#null-conditional-operators--and-) can simplify this block.

Using the null conditional allows you to change

```c#
            foreach (var componentController in ComponentControllers)
            {
                if (componentController == null)
                {
                    continue;
                }
                componentController.OnValidate();
            }
```

to 


```c#
            foreach (var componentController in ComponentControllers)
            {
                componentController?.OnValidate();
            }
```

NoelStephensUnity replied:
Good catch! _(will push that change shortly)_

EmandM commented on com.unity.netcode.gameobjects/Runtime/Components/Helpers/AttachableBehaviour.cs:
```diff
@@ -0,0 +1,572 @@
+        /// <summary>
+        /// The current <see cref="AttachableBehaviour"/> instance's <see cref="AttachState"/>.
+        /// </summary>
+        protected AttachState m_AttachState { get; private set; }
+
+        /// <summary>
+        /// The original parent of this <see cref="AttachableBehaviour"/> instance.
+        /// </summary>
+        protected GameObject m_DefaultParent { get; private set; }
+
+        /// <summary>
+        /// If attached, attaching, or detaching this will be the <see cref="AttachableNode"/> this <see cref="AttachableBehaviour"/> instance is attached to.
+        /// </summary>
+        protected AttachableNode m_AttachableNode { get; private set; }
```
Out of interest, why do these need to be protected rather than internal?

NoelStephensUnity replied:
This is so users have access to these properties if they chose to create a custom derived class of `AttachableBehaviour`.
Knowing:
- The current attached state is useful for an attachable.
- The current `AttachableNode` the custom `AttachableBehaviour` is attached to can be useful as well.
- The original parent of the `AttachableBehaviour` (i.e. the `GameObject` of its spawn-assigned `NetworkObject`) can be useful when the `AttachableBehaviour` is attached to something else and you need to get some property/field from that `GameObject` or any component (or nested component).

It is also why it is protected get and private set.

EmandM commented on com.unity.netcode.gameobjects/Runtime/Components/Helpers/AttachableBehaviour.cs:
```diff
@@ -0,0 +1,572 @@
+        /// If attached, attaching, or detaching this will be the <see cref="AttachableNode"/> this <see cref="AttachableBehaviour"/> instance is attached to.
+        /// </summary>
+        protected AttachableNode m_AttachableNode { get; private set; }
+        internal AttachableNode AttachableNode => m_AttachableNode;
```
Also out of interest, what is the value of this syntax rather than using `m_AttachableNode` directly?

NoelStephensUnity replied:
That is for internal access to the AttachableNode. I would have to go look, but it could very well be for testing purposes as well.

NoelStephensUnity replied:
Yeah,

It is used within [AttachableNode](https://github.com/Unity-Technologies/com.unity.netcode.gameobjects/blob/44c5ee79c59231e08b48daaccfea80dbe51e4a68/com.unity.netcode.gameobjects/Runtime/Components/Helpers/AttachableNode.cs#L54) and a few places within the `AttachableBehaviourTests`.

So, it has uses during normal runtime (via `AttachableNode`) and is used for integration testing.

NoelStephensUnity committed:
docs (update & add)

Adding `AttachableNode` and `ComponentController` sections.
Added "Synchronized RPC driven fields" section to networkbehaviour-synchronize.md.
Updated various areas based on the additions.

EmandM commented on com.unity.netcode.gameobjects/Runtime/Components/Helpers/AttachableBehaviour.cs:
```diff
@@ -0,0 +1,572 @@
+
+            // Exit early if we are already in the correct attached state and the incoming
+            // AttachableNode reference is the same as the local AttachableNode property.
+            if (attachableNode == m_AttachableNode &&
```
What happens if both `attachableNode` and `m_AttachableNode` are null here?

NoelStephensUnity replied:
Then that portion of the logical `if` check will be considered true. If the second portion of the `or` portion of that logical check is true then it will exit early.

EmandM commented on com.unity.netcode.gameobjects/Runtime/Components/Helpers/AttachableBehaviour.cs:
```diff
@@ -0,0 +1,572 @@
+            if (isAttaching && m_AttachableNode != null && m_AttachState == AttachState.Attached)
+            {
+                // If we are attached to some other AttachableNode, then detach from that before attaching to a new one.
+                if (m_AttachableNode != attachableNode)
```
I thiiiink, this if check will always be true. If it's false, I think it would have exited early on the block on line 300

NoelStephensUnity replied:
Just above that logical `if` check there is this line of code:
`isAttaching = isAttaching && attachableNode != null;`

If isAttaching is `false` when it reaches line 300, then it passes through to the next portion of script which would then set `isAttaching` to `true`. This check is primarily to only force `isAttaching` to `false` if the attachableNode is still `null`.

NoelStephensUnity replied:
_As a side note to this answer:_ 
You mentioned some concerns about the flow and dependency on properties. All of this seemingly strange logic handles what I am hoping are close to (if not all of) the majority of edge case scenarios that had high user pain potential. When looking at my responses and the code a second time, keep in mind that it handles the edge cases where the attach target `NetworkObject` (with one or more `AttachableNode` components) and/or the attachable `NetworkObject` (with one or more `AttachableBehaviour` components) are suddenly de-spawned due to a client ungracefully disconnecting. To handle these types of scenarios, the `AttachableBehaviour` and `ComponentController` have embedded logic to automatically adjust to the default without any need to detect the disconnection event.  
Look at:
- The [NetworkManager update](https://github.com/Unity-Technologies/com.unity.netcode.gameobjects/blob/44c5ee79c59231e08b48daaccfea80dbe51e4a68/com.unity.netcode.gameobjects/Runtime/Core/NetworkManager.cs#L1569).
  - The comments explain the why, but this adjustment of when things are despawned during shutdown help to cover a chunk of edge cases while also doubling up as a way to assure (if there is still a connection) any final state updates can be sent by the disconnecting client.
-  This portion of the [NetworkObject update](https://github.com/Unity-Technologies/com.unity.netcode.gameobjects/blob/44c5ee79c59231e08b48daaccfea80dbe51e4a68/com.unity.netcode.gameobjects/Runtime/Core/NetworkObject.cs#L2610)
  - This provides users with a "heads up... you are shutting down" notification on a per `NetworkBehaviour` instance basis.  If the transport connection is still valid during this time (a few scenarios where this will be true), then it is an opportunity to push the last final messages/state updates to the outbound queue.

All of these adjustments play into a lot of the quirky spots you might run into with `AttachableBehaviour`, `AttachableNode`, and `ComponentController`.

EmandM commented on com.unity.netcode.gameobjects/Runtime/Components/Helpers/AttachableBehaviour.cs:
```diff
@@ -0,0 +1,572 @@
+            }
+
+            var preNode = isAttaching ? attachableNode : m_AttachableNode;
+            isAttaching = isAttaching && attachableNode != null;
```
nit: Can we use two separate variables rather than reassigning this one? `hasReference` for the first block maybe? And the `isAttaching` from there?

NoelStephensUnity replied:
There really is no reason to do that. It is perfectly acceptable to use a field value to determine if the field value should change as long as there are other field, property, parameter, or some other local variable value calculated from other values is/are included in the check.

The use case is that even if the field is set to `true` we will set it to `false` if the `attachableNode` is `null` (and only under that one scenario).  The `isAttaching` is then used throughout the rest of the `UpdateAttachedState` method.  Adding yet another local variable would cause a bunch of additional code to be written and special case checks for the one scenario.

NoelStephensUnity committed:
docs (style - PVP)

Resolving more white space issues.

NoelStephensUnity committed:
Merge branch 'develop-2.0.0' into feat/attachable-networkbehaviour-and-object-controller

EmandM commented on com.unity.netcode.gameobjects/Runtime/Components/Helpers/AttachableBehaviour.cs:
```diff
@@ -0,0 +1,572 @@
+                }
+
+                // Developer only notification for the most likely scenario where this method is invoked but the instance is not attached to anything.
+                if (!m_AttachableNode && NetworkManager && NetworkManager.LogLevel <= LogLevel.Developer)
```
very nit: This if check can be put inside the block above to reduce the double check on `!m_AttachableNode`. The below check can then be swapped to an `else`, though I'm not sure if that makes any difference

NoelStephensUnity committed:
docs (style)

Removing a single sneaky white space.

EmandM commented on com.unity.netcode.gameobjects/Runtime/Components/Helpers/AttachableBehaviour.cs:
```diff
@@ -0,0 +1,572 @@
+        /// to have the owner be the authority of this <see cref="ComponentController"/> instance.
+        /// </remarks>
+        /// <returns>true = has authoriy | false = does not have authority</returns>
+        protected virtual bool OnHasAuthority()
```
I think there's a design conversation to be had here because we have [MTT-12442](https://jira.unity3d.com/browse/MTT-12442) open for this pattern on the `NetworkTransform`. We should at the very least align the naming with the `NetworkTransform` approach.

NoelStephensUnity replied:
Yeah... it was the only way to provide users with a way to make a client-server owner authoritative `AttachableBehaviour` while also providing users with a way to change the authority outside of our client-server and distributed authority restrictions.

I could think of a few ways where it would be useful to users to be able to override this specifically for "things that can be attached and detached" from something else...but for sure it is a bit clunky since everything else was designed well before this PR.

EmandM commented on com.unity.netcode.gameobjects/Runtime/Components/Helpers/AttachableBehaviour.cs:
```diff
@@ -0,0 +1,572 @@
+            return HasAuthority;
+        }
+
+        private void ChangeReference(NetworkBehaviourReference networkBehaviourReference)
```
I have a bit of a concern with the flow of 

`[Attach/Detatch]() -> ChangeReference() -> UpdateAttachedState() -> Internal[Attach/Detatch]()`

I'm uncomfortable with so much of the state of this flow going through the `m_` properties. I feel like we're going to end up with race conditions where one of those properties are updated unexpectedly. I'd feel more comfortable if the attach/detatch info was more explicitly and directly passed through the callstack, and/or the callstack was simplified some.

NoelStephensUnity replied:
This is part of a bigger discussion on "core" components and "helper" components. 

> I'd feel more comfortable if the attach/detatch info was more explicitly and directly passed through the callstack, and/or the callstack was simplified some.

This flow is actually much better the way it is put together currently using the synchronize RPC driven fields approach. Previously I was using NetworkVariable which that flow was much more complex and it did end up in certain race conditions.

This flow depends on the reliable sequenced ordering to assure that any sequence of events that changes the attached state from the start of the frame to the end of the frame will be received and in the order they happened during the frame.

EmandM commented on com.unity.netcode.gameobjects/Runtime/Components/Helpers/ComponentController.cs:
```diff
@@ -0,0 +1,543 @@
+                    // and use that as the time to add this next state update.
+                    if (m_PendingStateUpdates.Count > 0)
+                    {
+                        relativeTimeOffset = m_PendingStateUpdates[m_PendingStateUpdates.Count - 1].DelayTimeDelta;
```
Fun fact! C# has an "index from end" expression ([docs](https://learn.microsoft.com/en-us/dotnet/csharp/language-reference/proposals/csharp-8.0/ranges#systemindex)).

It could change

```csharp
    relativeTimeOffset = m_PendingStateUpdates[m_PendingStateUpdates.Count - 1].DelayTimeDelta;
```
to

```csharp
    relativeTimeOffset = m_PendingStateUpdates[^1].DelayTimeDelta;
```

More a comment for your info 😅. The current way is pretty readable rather than needing to remember new syntax 🤷‍♀️

NoelStephensUnity replied:
>  The current way is pretty readable rather than needing to remember new syntax

Oh for sure... I never like that addition as ^ is typically used as an XOR Boolean operation...which then forces you to stop and look at the context and then realize what it is doing.

I think we should stick to the more readable/traditional way.

EmandM commented on com.unity.netcode.gameobjects/Runtime/Components/Helpers/ComponentController.cs:
```diff
@@ -0,0 +1,543 @@
+        /// <summary>
+        /// Determines whether the selected <see cref="Components"/>s will start enabled or disabled when spawned.
+        /// </summary>
+        [Tooltip("The initial state of the component controllers enabled status when instnatiated.")]
```
```suggestion
        [Tooltip("The initial state of the component controllers enabled status when instantiated.")]
```

NoelStephensUnity replied:
🥇

EmandM commented on com.unity.netcode.gameobjects/Runtime/Components/Helpers/ComponentController.cs:
```diff
@@ -0,0 +1,543 @@
+                }
+
+                var propertyInfo = Components[i].Component.GetType().GetProperty("enabled", BindingFlags.Instance | BindingFlags.Public);
+                if (propertyInfo == null && propertyInfo.PropertyType != typeof(bool))
```
I'm guessing we don't want to check for `is null && has property`?

NoelStephensUnity replied:
Actually... that should be:
`if (propertyInfo == null || propertyInfo.PropertyType != typeof(bool))`

Thank you for commenting on that one! 
🥇

EmandM commented on com.unity.netcode.gameobjects/Runtime/Components/Helpers/ComponentController.cs:
```diff
@@ -0,0 +1,543 @@
+            // Scan the GameObject and all of its children and add all valid components to the list.
+            foreach (var entry in gameObjectsToScan)
+            {
+                var asGameObject = entry.Component as GameObject;
```
Does it make sense to change `gameObjectsToScan` from `List<ComponentEntry>` to `List<GameObject>`?

NoelStephensUnity replied:
So, that is actually a "side feature" for users where they may have so many components to add it might be faster to just "add all" components of a `GameObject` (and its children) and then remove the ones they might not want to include. Here is a quick video of this in action:


https://github.com/user-attachments/assets/67b5bb7b-7237-474d-a488-85480572f5a3

We want to keep that as a list of `GameObject`s.

EmandM replied:
Thanks for the explanation! 

I'm seeing that `gameObjectsToScan` is created as `List<ComponentEntry>` ([line 226](https://github.com/Unity-Technologies/com.unity.netcode.gameobjects/pull/3518/files/fe889232aa97af99528452eebaf177b9919e5c00#diff-c8dda3f44ea07cbcb7f349be0db86ec4812b54eeae5adc40046c73151d09082aR226)) and then the type of the component is checked on [lines 239 & 240](https://github.com/Unity-Technologies/com.unity.netcode.gameobjects/pull/3518/files/fe889232aa97af99528452eebaf177b9919e5c00#diff-c8dda3f44ea07cbcb7f349be0db86ec4812b54eeae5adc40046c73151d09082aR239-R240). The Component is then cast into a `GameObject` here. I'm wondering if it makes more sense to simply build a list of the `GameObject`s rather than essentially doing the cast twice.

NoelStephensUnity replied:
It only casts the entry.Component to a `GameObject` once... ??

It checks if the `Components[i].Component` type is of type `GameObject` and if so it adds the existing class instance to the gameObjectsToScan list.

Then, while processing through the gameObjectsToScan list, it does the single cast from `UnityEngine.Object` to `UnityEngine.GameObject` in order to just get the components.

Once it has the components of the `GameObject`, it does't really need that entry any longer.

EmandM commented on com.unity.netcode.gameobjects/Runtime/Components/Helpers/ComponentController.cs:
```diff
@@ -0,0 +1,543 @@
+            gameObjectsToScan.Clear();
+
+            // Final (third) pass is to name each list element item as the component is normally viewed in the inspector view.
+            for (int i = 0; i < Components.Count; i++)
```
very nit: This might be simpler to read as a `foreach` loop. I do not have strong feelings about this.

NoelStephensUnity replied:
No worries... I can do that for sure...

EmandM commented on com.unity.netcode.gameobjects/Runtime/Components/Helpers/ComponentController.cs:
```diff
@@ -0,0 +1,543 @@
+        /// <inheritdoc/>
+        protected override void OnSynchronize<T>(ref BufferSerializer<T> serializer)
+        {
+            // Example of how to synchronize late joining clients when using an RPC to update
+            // a local property's state.
```
Is this comment valid here?

NoelStephensUnity replied:
Yes it is... 
There is this new section on [Synchronized RPC driven fields](https://github.com/Unity-Technologies/com.unity.netcode.gameobjects/blob/5662003a0aa4c3ad4456a66b54e13182fddeb53e/com.unity.netcode.gameobjects/Documentation%7E/components/foundational/networkbehaviour-synchronize.md#synchronized-rpc-driven-fields) that explains what the above is referring to.

Actually, let me know if after reading the above added sub-section the comment makes more sense... it will be a good metric on whether that does a good job explaining things or not.

EmandM replied:
Ah! That makes sense. I should have read those docs 😅

EmandM commented on com.unity.netcode.gameobjects/Runtime/Components/Helpers/ComponentController.cs:
```diff
@@ -0,0 +1,543 @@
+        /// to have the owner be the authority of this <see cref="ComponentController"/> instance.
+        /// </remarks>
+        /// <returns>true = has authoriy | false = does not have authority</returns>
+        protected virtual bool OnHasAuthority()
```
Same comment as for `AttachableBehaviour` here

NoelStephensUnity replied:
So....
Originally this was added so a user could make the `ComponentController` "owner authoritative" when using a client-server network topology (_i.e. HasAuthority only returns true on server instances_). However, I then realized that using the synchronized RPC driven approach provides users with a way for "anyone" to be the "authority" (i.e. the one who can toggle the enabled state)... which is why the RPC is using the `SendTo.NotMe` target...

Use case:
A user makes a game where there are teams and the user creates an asset where some in-scene placed Flag will have different "views" for each team and will "transform" to a different visual state when it is picked up or interacted with... with this method a user could determine which `NetworkManager`/client instance is trying to interact with the `ComponentController` and if the client is on the same team that is allowed to make changes then the user can return `true`.

NoelStephensUnity committed:
Update com.unity.netcode.gameobjects/Runtime/Core/NetworkObject.cs

Co-authored-by: Emma <emma.mcmillan@unity3d.com>

NoelStephensUnity committed:
update

Implementing some suggested changes from PR review.

NoelStephensUnity committed:
style

Adding whitespace.

NoelStephensUnity committed:
Merge branch 'develop-2.0.0' into feat/attachable-networkbehaviour-and-object-controller

jabbacakes committed:
Lorge doc review

jabbacakes commented:
Okay I just pushed a _large_ docs commit, hear me out:

- The only changes that need in-depth review are those to the three new helper components, it's mostly just tweaking wording, removing 'we' and other colloquialisms that are no bueno in the docs.
- I also did a bunch of mass changes to convert all component names into regular, non-monospace. We've been inconsistent about that for a while, and I checked the [style guide](https://docs-style-guide.unity.com/terminology/terminology-list#c) and it said do it that way. So FYI but you can ignore those changes throughout the wider docs, I just kept bumping into the inconsistency and it was driving me barmy.
- Changed from foundational components -> core components, as discussed with Noel.

All in all, I think this is a really good body of doc to launch this feature with, and we can finesse it if/when we get user feedback about it.

NoelStephensUnity committed:
Merge branch 'develop-2.0.0' into feat/attachable-networkbehaviour-and-object-controller

NoelStephensUnity committed:
Merge branch 'develop-2.0.0' into feat/attachable-networkbehaviour-and-object-controller

NoelStephensUnity committed:
Update AttachableBehaviour.cs

Applying modifications suggested by Emma.

NoelStephensUnity committed:
style

Remove whitespace

NoelStephensUnity committed:
Update com.unity.netcode.gameobjects/Runtime/Components/Helpers/ComponentController.cs

Co-authored-by: Emma <emma.mcmillan@unity3d.com>

NoelStephensUnity committed:
Merge branch 'develop-2.0.0' into feat/attachable-networkbehaviour-and-object-controller

NoelStephensUnity committed:
Merge branch 'develop-2.0.0' into feat/attachable-networkbehaviour-and-object-controller

NoelStephensUnity committed:
update

Modifications from 1:1 PR review session with Emma.
Removed some redundant logic.
Re-organized a few areas for better readability.
Renamed the `AttachableBehaviour.AttachableNode` to  `AttachableBehaviour.InternalAttachableNode`.
Removed the authority check within `AttachableNode.OnNetworkPreDespawn`.

NoelStephensUnity committed:
update

Final suggested modifications from 1:1 with Emma during PR review.

EmandM reviewed (approved):
:godmode: 🥇

NoelStephensUnity committed:
doc: update

Adding OnNetworkPreDespawn documentation.
Adjusted the NetworkBehaviour documentation to include some best practices information.

jabbacakes commented on com.unity.netcode.gameobjects/Documentation~/components/core/networkbehaviour.md:
```diff
@@ -79,49 +140,56 @@ public class MyNetworkBehaviour : NetworkBehaviour
-### Destroying
+### De-spawning but not destroying
+
+ When de-spawned and not destroyed, the associated `GameObject` instance, and all children of that `GameObject`, will persist until it is destroyed. Under this scenario (_de-spawn but not destroy_), the would de-spawn but not destroy the `NetworkObject` instance with the intention of re-using/re-spawning the instance. In order to de-spawn and not destroy, you must invoke `NetworkObject.Despawn` w ...
```
'When de-spawned and not destroyed, the associated `GameObject` instance, and all children of that `GameObject`, will persist until it is destroyed.'

What does 'it' refer to in this sentence? Is it the GameObject or the NetworkObject?

NoelStephensUnity committed:
Merge branch 'develop-2.0.0' into feat/attachable-networkbehaviour-and-object-controller

jabbacakes committed:
Follow up docs pass

netcode-ci-service committed:
Merge develop-2.0.0 into feat/attachable-networkbehaviour-and-object-controller

NoelStephensUnity committed:
Merge branch 'develop-2.0.0' into feat/attachable-networkbehaviour-and-object-controller

Mechitary commented:
thanks Noel, it looks like these components will alleviate a lot of unneeded complexity. 

With the `AttachableBehaviour` component, can I parent a network animator that has no network object (i.e. a prefab with its own network animator) to an already spawned network object? Current it does not work without a network object and network animator within the same hierarchy (i.e. prefab), due to how network animator is initialized in its `Awake`.

mech
